docs/src/developer/developer/cassandra-interaction.md

```diff
@@ -1,46 +1,301 @@
+
+### Background
+
+We currently operate all our production and production-like environments with
+
+* a **replication factor of 3** (per datacentre, but except for a brief phase of controlled
+  datacentre migration there is only one datacentre in use)
+* a number of cassandra nodes >= 3 (depending on load and disk/machine size)
+* Everywhere where some consistency matters, **Quorum writes and Quorum reads**
+
+These three things together give us **strong consistency on a per-table** basis. A quorum write
+followed by a quorum read will guarantee that you see the data you inserted, even if one replica of
+cassandra that holds that partition key dies in the middle.
+
+The replication factor is specified when creating or migrating schemas, which is done in the
+`cassandra-migrations` subchart of the `wire-server` chart:
+
+```{grepinclude} ../charts/cassandra-migrations/values.yaml host name and replication
+---
+lines-after: 6
+language: yaml
+---
+```
+
+The number of cassandra nodes in use is specified on the infrastructure level (k8ssandra on
+kubernetes, or the inventory list when using ansible-cassandra)
+
+Quorum consistency (or, in our case, `LocalQuorum` consistency) is specified in our code. Random
+example:
+
+```{grepinclude} ../services/brig/src/Brig/Data/User.hs userEmailUpdate \(params
+---
+language: Haskell
+---
+```
+
+Note that `Quorum` and `LocalQuorum` behave exactly the same in the context of a single datacentre
+(each datacentre can have multiple racks or availability zones). We switched from `Quorum` to
+`LocalQuorum` in [this PR](https://github.com/wireapp/wire-server/pull/1884) in preparation of a
+datacentre migration. As it doesn't hurt, and more datacentre migrations may be done in the future
+by us or the people hosting on-premise, let's stick to `LocalQuorum` for the time being.
+
+***To summarize: In case of doubt, just follow the examples and use `LocalQuorum` Writes and Reads
+in your queries, unless there is a very good reason not to.***
+
+(batch-statements)=
+## Guidelines for across-table consistency
+
+Given that we have strong consistency for a single operation (see section above), in some cases your
+data access patterns mandate that you have multiple tables with seemingly duplicated data.
+
+Good ✅ example: For instance, a 'teamInvitation' needs to be queried
+* by team ID (for the team admin- which pending invitations exist?)
+* by email (during user registration), and
+* by code (to check whether an invitation is valid before embarking on the user registration).
+
+Therefore, there are three tables holding similar data about a team invitation, but with different
+primary keys. This is a correct and recommended data design for cassandra.
+
+Now, any insert, update or delete of this data should either all work, or all fail, to not have
+weird data integrity issues. Now, cassandra provides for this, and it's called `batch` statements.
```

Comment:
This is another case of misleading terminology in datastax. Batches are actually **not** atomic as the A in ACID.

First, batches are not rolled back in case of failures, so realistically if something goes wrong mid-batch, you'll get a timeout, hence a failed request, even though the coordinator will probably retry the batch eventually and get it through.

Secondly, there is not even a guarantee of batches being eventually applied in full, in case the coordinator fails. See for example: https://www.datastax.com/blog/atomic-batches-cassandra-12.

Most importantly, though, batches are not **isolated** (as the I in ACID), in the sense that other processes can see partially applied changes. You touch on this later but say that "it's not a problem". I don't get why not.

Reply:
I added a note about the lack of isolation. I further simply claim that we can not care about it. This claim is not substantiated well, finding a counter-example is enough to destroy that claim. I invite you to find one in the Wire context that shows that at least in some cases we must care about it.

Meanwhile I think we should use batch statements more often, as, as indicated in the chat already, we don't do this a lot, and lack of any kind of atomicity introduces real data inconsistency based bugs that are very hard to fix, and super annoying for the users concerned.